File: backend/src/main/java/com/recruitpro/model/Job.java
package com.recruitpro.model;

import com.recruitpro.model.enums.ExperienceLevel;
import com.recruitpro.model.enums.JobStatus;
import com.recruitpro.model.enums.JobType;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.annotations.SQLRestriction;
import org.hibernate.type.SqlTypes;

import java.time.Instant;
import java.time.LocalDate;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.UUID;

@Entity
@Table(name = "jobs")
@SQLRestriction("deleted_at IS NULL")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Job {

    @Id
    @Column(updatable = false, nullable = false)
    private UUID id;

    @Column(name = "company_id", nullable = false)
    private UUID companyId;

    @Column(name = "company_address_id")
    private UUID companyAddressId;

    @Column(nullable = false)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "industry_id")
    private Industry industry;

    @Column(name = "responsibilities", columnDefinition = "text[]")
    @JdbcTypeCode(SqlTypes.ARRAY)
    private String[] responsibilities;

    @Column(name = "requirements", columnDefinition = "text[]")
    @JdbcTypeCode(SqlTypes.ARRAY)
    private String[] requirements;

    @Column(name = "nice_to_have_skills", columnDefinition = "text[]")
    @JdbcTypeCode(SqlTypes.ARRAY)
    private String[] niceToHaveSkills;

    @Column(name = "job_data_structure", columnDefinition = "jsonb")
    @JdbcTypeCode(SqlTypes.JSON)
    private Map<String, Object> jobDataStructure;

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(
        name = "job_experience_levels",
        joinColumns = @JoinColumn(name = "job_id")
    )
    @Column(name = "level", columnDefinition = "experience_level")
    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Builder.Default
    private Set<ExperienceLevel> experienceLevels = new HashSet<>();

    private String location;

    @Column(name = "salary_min")
    private Integer salaryMin;

    @Column(name = "salary_max")
    private Integer salaryMax;

    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Column(name = "job_type", columnDefinition = "job_type")
    private JobType jobType;

    @Enumerated(EnumType.STRING)
    @JdbcTypeCode(SqlTypes.NAMED_ENUM)
    @Column(nullable = false, columnDefinition = "job_status")
    private JobStatus status;

    @Column(name = "close_date")
    private LocalDate closeDate;

    // embedding is managed by AI service — not mapped as JPA column to avoid pgvector deps in JPA
    // Use @Column(columnDefinition = "vector(384)") when pgvector JPA support is added

    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
        name = "job_skills",
        joinColumns = @JoinColumn(name = "job_id"),
        inverseJoinColumns = @JoinColumn(name = "skill_id")
    )
    @Builder.Default
    private Set<Skill> skills = new HashSet<>();

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "updated_at")
    private Instant updatedAt;

    @Column(name = "deleted_at")
    private Instant deletedAt;

    @Column(name = "attachment_url", length = 1024)
    private String attachmentUrl;

    @Transient
    private long applicationCount;

    @PrePersist
    protected void onCreate() {
        if (id == null) id = UUID.randomUUID();
        if (status == null) status = JobStatus.DRAFT;
        createdAt = Instant.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = Instant.now();
    }
}

File: frontend-company/src/types/job.ts
export type JobStatus = 'DRAFT' | 'PUBLISHED' | 'CLOSED';
export type JobType = 'FULLTIME' | 'PARTTIME' | 'REMOTE' | 'HYBRID';
export type ExperienceLevel = 'INTERN' | 'FRESHER' | 'JUNIOR' | 'MIDDLE' | 'SENIOR' | 'LEADER';

export interface Industry {
  id: string;
  name: string;
}

export interface Job {
  id: string;
  title: string;
  description: string;
  industry: Industry | null;
  responsibilities: string[];
  requirements: string[];
  jobType: JobType;
  experienceLevels: ExperienceLevel[];
  location: string;
  salaryMin: number | null;
  salaryMax: number | null;
  status: JobStatus;
  closeDate?: string | null;
  skills: Skill[];
  niceToHaveSkills: string[];
  applicationCount: number;
  createdAt: string;
  updatedAt: string;
  attachmentUrl?: string;
  companyAddressId?: string;
}

export interface JobFormData {
  title: string;
  description: string;
  industryId?: string;
  responsibilities: string[];
  requirements: string[];
  jobType: JobType;
  levels: ExperienceLevel[];
  location: string;
  addressId?: string;
  salaryMin?: number;
  salaryMax?: number;
  skillIds: string[];
  niceToHaveSkills: string[];
  status: JobStatus;
  closeDate?: string | null;
  attachmentUrl?: string;
}

export interface JobSelectOption {
  id: string;
  title: string;
}

export interface Skill {
  id: string;
  name: string;
}

export interface JobAutoFillDto {
  jobTitle?: string;
  industry?: string;
  jobType?: string;
  experienceLevels?: string[];
  location?: string;
  salaryMin?: number;
  salaryMax?: number;
  skills?: string[];
  responsibilities?: string[];
  requirements?: string[];
  niceToHaveSkills?: string[];
  description?: string;
  mustHaveSkills?: string[];
}

export type JobAutoFillApiResponse = JobAutoFillDto;
